### DISCUSSION TITLE
Using additional buttons without affecting rotary menu

### DISCUSSION THREAD

**Discussion by acabrera2012** [Spirik/GEM#85] [Q&A] [answered]
Hello.

I have a need to add 2 buttons to a project that are not associated with the rotary encoder or any part of the menus. The 2 buttons just need to execute some function separately. When I try to use these simple buttons either by polling their state such as digitalRead(GREEN_BUTTON_PIN) or by attaching an interrupt such as attachInterrupt(digitalPinToInterrupt(GREEN_BUTTON_PIN), checkGreenButton, FALLING); the rotary menu stops working.

Here is a link to the project. If you uncomment the red or green button code it causes issues with the rotary functionality : [https://wokwi.com/projects/392458742047807489](url)

Any advice would be greatly appreciated.

**Comment by Spirik** [comment]
Hello, @acabrera2012!

Thank you for providing Wokwi implementation! Indeed, it seems that rotary encoder button is not working reliably for some reason (regardless of presence of another buttons). I will look into it.

However, there are couple of things I've noticed at a quick glance: in Wokwi environment it is better to set `debounceDelay` of KeyDerector object  to 0 (since Wokwi is very limited in simulation of physical effects, such as signal ripple, and analog circuitry) and use `U8G2_SSD1306_128X64_NONAME_1_HW_I2C` U8g2 class (instead of `U8G2_SH1106_128X64_NONAME_1_HW_I2C` - they are similar, however some undesirable shift in screen space coordinates is present resulting in cropping of image to the right of the screen).

Will investigate further.

**Reply by acabrera2012** [reply]
Thanks for getting back to me, Alexander. Great job with this library. I really want to use it but I do need to be able to use those 2 other buttons without interfering with the rotary menu.

I made the changes you recommended above but now the rotary push button is not working. It does seem to matter which pin I attach to the SW on the rotary encoder. By the way, I am using a Nano and I know it only has 2 interrupt pins (2, 3)  but I don't think that's what's causing my pain.

Thanks for looking into it. Let me know if you find anything out. Hopefully, it's just something dumb I did on my end.

Thanks again.

**Reply by Spirik** [reply]
Curiously enough, version you've posted to Arduino forum has no problems with operation of rotary encoder, so it's possible that something happened in `processMenu()` during your experiments. I will try to combine your recent code (posted here) with initial implementation of `processMenu()` , which I will copy from Arduino forum listing (you may have successfully deleted it from forum, but I still have it in my email notification :) ). The route you've chosen in most recent version - to use KeyDetector to detect presses of additional buttons - seems like the most promising one so far.

**Comment by acabrera2012** [comment]
Sounds good. Let me know what you find.

Just to be sure, here is the link to the Wokwi project again: [https://wokwi.com/projects/392458742047807489](url).

Thanks again.

**Comment by Spirik** [comment, marked as answer]
Who would have thought, seems like a memory issue after all=)

As soon as I've changed all of `Serial.println()` calls to use [`F()`macro](https://www.arduino.cc/reference/en/language/variables/utilities/progmem/) everything works just fine.

Here's Wokwi: https://wokwi.com/projects/392813742679205889
Here's Gist: https://gist.github.com/Spirik/abf77ccd9cd309d67b5d8ead0599a06b

**Reply by acabrera2012** [reply]
Wow! You are the man. I'm sorry I wasted your time but it didn't even occur to me that it could be a memory issue. Maybe the Nano's limited resources couldn't handle it. I'm going to try it on some real hardware instead of Wokwi to see how it goes.

Thanks again!!!!

**Reply by acabrera2012** [reply]
...Just tried it on real Nano hardware and it works like a champ! Thanks so much!!!!!!!!!

**Reply by Spirik** [reply]
No problem! Glad I was able yo help!

PS: Do not forget to adjust `debounceDelay` for the real hardware=)

**Reply by acabrera2012** [reply]
Will do. Thanks
